File: ios/DrawingEngine/Sources/DrawingEngine/Fitting/ShapeFitter.swift
import Foundation

public struct FittedCircle: Sendable, Equatable {
    public let center: Point2D
    public let radius: Double
    /// RMS distance of the points from the fitted circle. Lower = rounder.
    public let rmsResidual: Double
}

public struct FittedLine: Sendable, Equatable {
    public let start: Point2D
    public let end: Point2D
    /// RMS perpendicular distance from the points to the fitted line. Lower = straighter.
    public let rmsPerpendicular: Double
}

/// Fits a freehand stroke to a primitive to extract shape-intrinsic quality
/// (roundness, straightness) independent of where on the canvas it was drawn.
public enum ShapeFitter {

    /// Algebraic (Kåsa) circle fit: solves x² + y² = A·x + B·y + C in the
    /// least-squares sense, then recovers center = (A/2, B/2), r = √(C + a² + b²).
    public static func fitCircle(_ points: [Point2D]) -> FittedCircle? {
        guard points.count >= 3 else { return nil }
        let n = Double(points.count)
        var sxx = 0.0, sxy = 0.0, syy = 0.0, sx = 0.0, sy = 0.0
        var sxz = 0.0, syz = 0.0, sz = 0.0
        for p in points {
            let z = p.x * p.x + p.y * p.y
            sxx += p.x * p.x
            sxy += p.x * p.y
            syy += p.y * p.y
            sx += p.x
            sy += p.y
            sxz += p.x * z
            syz += p.y * z
            sz += z
        }
        let matrix = [[sxx, sxy, sx],
                      [sxy, syy, sy],
                      [sx,  sy,  n]]
        let rhs = [sxz, syz, sz]
        guard let solution = LinearAlgebra.solve3x3(matrix, rhs) else { return nil }

        let a = solution[0] / 2
        let b = solution[1] / 2
        let r2 = solution[2] + a * a + b * b
        guard r2 > 0 else { return nil }
        let radius = r2.squareRoot()
        let center = Point2D(a, b)

        var sumSq = 0.0
        for p in points {
            let d = p.distance(to: center) - radius
            sumSq += d * d
        }
        let rms = (sumSq / n).squareRoot()
        return FittedCircle(center: center, radius: radius, rmsResidual: rms)
    }

    /// Total-least-squares line fit via PCA: the principal eigenvector of the
    /// point covariance is the line direction; endpoints are the extreme projections.
    public static func fitLine(_ points: [Point2D]) -> FittedLine? {
        guard points.count >= 2 else { return nil }
        let n = Double(points.count)
        let centroid = Geometry.centroid(points)

        var cxx = 0.0, cxy = 0.0, cyy = 0.0
        for p in points {
            let dx = p.x - centroid.x
            let dy = p.y - centroid.y
            cxx += dx * dx
            cxy += dx * dy
            cyy += dy * dy
        }
        cxx /= n; cxy /= n; cyy /= n

        let trace = cxx + cyy
        let det = cxx * cyy - cxy * cxy
        let disc = max(0, trace * trace / 4 - det)
        let lambda = trace / 2 + disc.squareRoot()   // larger eigenvalue

        var direction: Point2D
        if abs(cxy) > 1e-12 {
            direction = Point2D(lambda - cyy, cxy)
        } else {
            direction = cxx >= cyy ? Point2D(1, 0) : Point2D(0, 1)
        }
        let dlen = direction.length
        direction = dlen > 0 ? Point2D(direction.x / dlen, direction.y / dlen) : Point2D(1, 0)
        let normal = Point2D(-direction.y, direction.x)

        var tMin = Double.infinity, tMax = -Double.infinity, sumPerpSq = 0.0
        for p in points {
            let v = p - centroid
            let t = v.dot(direction)
            tMin = min(tMin, t)
            tMax = max(tMax, t)
            let perp = v.dot(normal)
            sumPerpSq += perp * perp
        }
        let start = Point2D(centroid.x + direction.x * tMin, centroid.y + direction.y * tMin)
        let end = Point2D(centroid.x + direction.x * tMax, centroid.y + direction.y * tMax)
        let rms = (sumPerpSq / n).squareRoot()
        return FittedLine(start: start, end: end, rmsPerpendicular: rms)
    }
}

File: web/src/lib/engine/shapeFitter.ts
import type { Point2D } from './geometry'
import { centroid, distance } from './geometry'
import { solve3x3 } from './linearAlgebra'

export interface FittedCircle {
  readonly center: Point2D
  readonly radius: number
  /** RMS distance from points to the fitted circle. Lower = rounder. */
  readonly rmsResidual: number
}

export interface FittedLine {
  readonly start: Point2D
  readonly end: Point2D
  /** RMS perpendicular distance from points to the fitted line. Lower = straighter. */
  readonly rmsPerpendicular: number
}

/**
 * Algebraic (Kåsa) circle fit: solve x² + y² = A·x + B·y + C in the least-squares
 * sense, then recover center = (A/2, B/2), r = √(C + a² + b²).
 */
export function fitCircle(points: readonly Point2D[]): FittedCircle | null {
  if (points.length < 3) return null
  const n = points.length
  let sxx = 0, sxy = 0, syy = 0, sx = 0, sy = 0
  let sxz = 0, syz = 0, sz = 0
  for (const p of points) {
    const z = p.x * p.x + p.y * p.y
    sxx += p.x * p.x; sxy += p.x * p.y; syy += p.y * p.y
    sx += p.x; sy += p.y
    sxz += p.x * z; syz += p.y * z; sz += z
  }
  const matrix = [
    [sxx, sxy, sx],
    [sxy, syy, sy],
    [sx, sy, n]
  ]
  const sol = solve3x3(matrix, [sxz, syz, sz])
  if (!sol) return null
  const a = sol[0] / 2
  const b = sol[1] / 2
  const r2 = sol[2] + a * a + b * b
  if (r2 <= 0) return null
  const radius = Math.sqrt(r2)
  const center: Point2D = { x: a, y: b }
  let sumSq = 0
  for (const p of points) {
    const d = distance(p, center) - radius
    sumSq += d * d
  }
  return { center, radius, rmsResidual: Math.sqrt(sumSq / n) }
}

/** Total-least-squares line fit via PCA. */
export function fitLine(points: readonly Point2D[]): FittedLine | null {
  if (points.length < 2) return null
  const n = points.length
  const c = centroid(points)
  let cxx = 0, cxy = 0, cyy = 0
  for (const p of points) {
    const dx = p.x - c.x, dy = p.y - c.y
    cxx += dx * dx; cxy += dx * dy; cyy += dy * dy
  }
  cxx /= n; cxy /= n; cyy /= n
  const trace = cxx + cyy
  const det = cxx * cyy - cxy * cxy
  const disc = Math.max(0, (trace * trace) / 4 - det)
  const lambda = trace / 2 + Math.sqrt(disc)
  let dx: number, dy: number
  if (Math.abs(cxy) > 1e-12) {
    dx = lambda - cyy
    dy = cxy
  } else if (cxx >= cyy) {
    dx = 1; dy = 0
  } else {
    dx = 0; dy = 1
  }
  const dlen = Math.hypot(dx, dy)
  if (dlen > 0) { dx /= dlen; dy /= dlen } else { dx = 1; dy = 0 }
  const nx = -dy, ny = dx
  let tMin = Infinity, tMax = -Infinity, sumPerpSq = 0
  for (const p of points) {
    const vx = p.x - c.x, vy = p.y - c.y
    const t = vx * dx + vy * dy
    if (t < tMin) tMin = t
    if (t > tMax) tMax = t
    const perp = vx * nx + vy * ny
    sumPerpSq += perp * perp
  }
  const start: Point2D = { x: c.x + dx * tMin, y: c.y + dy * tMin }
  const end: Point2D = { x: c.x + dx * tMax, y: c.y + dy * tMax }
  return { start, end, rmsPerpendicular: Math.sqrt(sumPerpSq / n) }
}
